src/shogun/classifier/LDA.cpp

```diff
@@ -91,59 +93,81 @@ bool CLDA::train_machine(CFeatures *data)
 
 template <typename ST>
 bool CLDA::train_machine_templated(SGVector<int32_t> train_labels, CFeatures *data)
-{	
-	SGMatrix<ST>feature_matrix=((CDenseFeatures<ST>*)features)
-										->get_feature_matrix();
+{
+	SGMatrix<ST> feature_matrix =
+	    ((CDenseFeatures<ST>*)features)->get_feature_matrix();
 	int32_t num_feat=feature_matrix.num_rows;
 	int32_t num_vec=feature_matrix.num_cols;
 
-	SGVector<int32_t> classidx_neg(num_vec);
-	SGVector<int32_t> classidx_pos(num_vec);
-
-	int32_t i=0;
-	int32_t num_neg=0;
-	int32_t num_pos=0;
-
-	for(i=0; i<train_labels.vlen; i++)
+	if (m_method == SVD_LDA || (m_method == AUTO_LDA && num_vec <= num_feat))
 	{
-		if (train_labels.vector[i]==-1)
-			classidx_neg[num_neg++]=i;
-
-		else if(train_labels.vector[i]==+1)
-			classidx_pos[num_pos++]=i;
+		std::unique_ptr<CFisherLDA> lda(new CFisherLDA(CANVAR_FLDA));
+		std::unique_ptr<CMulticlassLabels> multiclass_labels(
+		    new CMulticlassLabels(m_labels->get_num_labels()));
+
+		for (index_t i = 0; i < m_labels->get_num_labels(); ++i)
+			multiclass_labels->set_int_label(
+			    i, (((CBinaryLabels*)m_labels)->get_int_label(i) == 1 ? 1 : 0));
+		lda->fit(data, multiclass_labels.get(), 1);
+
+		auto m = lda->get_transformation_matrix();
+		SGVector<float64_t> w(num_feat);
+		for (index_t i = 0; i < num_feat; ++i)
+			w[i] = m[i];
+
+		set_w(w);
+		set_bias(-linalg::dot(w, lda->get_mean()));
 	}
-
-	SGVector<ST> w_st(num_feat);
-	w_st.zero();
-	typename SGMatrix<ST>::EigenMatrixXt fmatrix=typename SGMatrix<ST>::EigenMatrixXtMap(feature_matrix.matrix, num_feat, num_vec);
-	typename SGVector<ST>::EigenVectorXt mean_neg(num_feat);
-	mean_neg.setZero();
-	typename SGVector<ST>::EigenVectorXt mean_pos(num_feat);
-	mean_pos.setZero();
-
-	//mean neg
-	for(i=0; i<num_neg; i++)
-		mean_neg+=fmatrix.col(classidx_neg[i]);
-	mean_neg/=(ST)num_neg;
-
-	// get m(-ve) - mean(-ve)
-	for(i=0; i<num_neg; i++)
-		fmatrix.col(classidx_neg[i])-=mean_neg;
-
-	//mean pos
-	for(i=0; i<num_pos; i++)
-		mean_pos+=fmatrix.col(classidx_pos[i]);
-	mean_pos/=(ST)num_pos;
-
-	// get m(+ve) - mean(+ve)
-	for(i=0; i<num_pos; i++)
-		fmatrix.col(classidx_pos[i])-=mean_pos;
-
-	SGMatrix<ST>scatter_matrix(num_feat, num_feat);
-	typename SGMatrix<ST>::EigenMatrixXtMap scatter(scatter_matrix.matrix, num_feat, num_feat);
-
-	if (m_method == FLD_LDA || (m_method==AUTO_LDA && num_vec>num_feat))
+	else
 	{
+		SGVector<ST> w_st(num_feat);
+
+		SGVector<int32_t> classidx_neg(num_vec);
+		SGVector<int32_t> classidx_pos(num_vec);
+
+		int32_t i = 0;
+		int32_t num_neg = 0;
+		int32_t num_pos = 0;
+
+		for (i = 0; i < train_labels.vlen; i++)
+		{
+			if (train_labels.vector[i] == -1)
+				classidx_neg[num_neg++] = i;
+
+			else if (train_labels.vector[i] == +1)
+				classidx_pos[num_pos++] = i;
+		}
+
+		typename SGMatrix<ST>::EigenMatrixXt fmatrix =
```

Comment:
the problem with these eigen mapped matricies that in case the linalg backend is on gpu these will all fail, as feature matrix is not on the CPU's mem. :(